Document the new JS API

nex3 opened the pull request:
This shouldn't be merged until the new API is supported by Dart Sass.

stof commented on js-api-doc/exception.d.ts:
```diff
@@ -1,13 +1,41 @@
+   * Unlike [[message]], this does *not* include representations of [[span]] or
+   * [[sassStack]].
+   */
   sassMessage: string;
```
@nex3 should those be typed as readonly properties ? AFAICT, dart-sass implements them as getters

nex3 replied:
Yeah, good call.

Awjin commented on js-api-doc/README.md:
````diff
@@ -1,15 +1,75 @@
+  ```js
+  const sass = require('sass');
+
+  const result = sass.compile("style.css");
````
```suggestion
  const result = sass.compile("style.scss");
```

Awjin commented on js-api-doc/value/string.d.ts:
```diff
@@ -1,21 +1,87 @@
+   * @throws `Error` if `sassIndex` isn't a number, if that number isn't an
+   * integer, or if that integer isn't a valid index for this string.
+   *
+   * @throws `Error` If `sassIndex` isn't a number, if that number isn't an
```
```suggestion
   * @throws `Error` if `sassIndex` isn't a number, if that number isn't an
```

Awjin commented on spec/js-api/value/boolean.d.ts:
```diff
@@ -7,6 +7,8 @@ export const sassTrue: SassBoolean;
 
 /** The JS API representation of a Sass boolean. */
-export interface SassBoolean extends Value {
+export class SassBoolean extends Value {
+  private constructor();
+
```
Why is this changed to a class?

nex3 replied:
To match other types, and so that `instanceof SassBoolean` works if necessary.

Awjin commented on js-api-doc/value/list.d.ts:
```diff
@@ -2,25 +2,57 @@ import {List} from 'immutable';
 import {Value} from './index';
 
+/**
+ * Possible separators used by Sass lists. The special separator `null` is only
+ * used for lists with fewer than two elements, and indicates that the separator
+ * has not yet been decided for this list.
```
Is this still true for the new API? (the constructor defaults to comma separator) I think this should be null only for scalar values

nex3 replied:
Yes, it just means that it's *possible* to use the `null` separator for single-element lists—any list may have a non-null separator, but only 0- or 1-element lists may have a null one.

Awjin commented on js-api-doc/value/list.d.ts:
```diff
@@ -2,25 +2,57 @@ import {List} from 'immutable';
     brackets?: boolean;
   }): SassList;
 
+  /** @hidden */
```
Does this need to stay hidden?

nex3 replied:
Similarly to `Value.toString()`, it doesn't provide any additional documentation beyond the superclass's method so it's better to just omit it. (I wish TypeDoc omitted undocumented overrides by default, but sadly it does not.)

Awjin commented on js-api-doc/value/map.d.ts:
```diff
@@ -2,12 +2,29 @@ import {OrderedMap} from 'immutable';
+   */
   get contents(): OrderedMap<Value, Value>;
 
+  /** @hidden */
```
Does this need to stay hidden?

Awjin commented on js-api-doc/value/index.d.ts:
```diff
@@ -17,40 +17,144 @@ export {SassMap} from './map';
+  /** Returns a hash code that can be used to store `this` in a hash map. */
   hashCode(): number;
 
+  /** @hidden */
```
Does this need to stay hidden?

nex3 replied:
It doesn't need to, but it's kind of unsightly in the generated documentation and doesn't really provide any additional information since `Object.toString()` already exists.

Awjin commented on js-api-doc/logger/source_span.d.ts:
```diff
@@ -2,14 +2,35 @@ import {URL} from 'url';
   start: SourceLocation;
 
+  /**
+   * The end of this span covers, exclusive.
```
```suggestion
   * The end of this span, exclusive.
```

Awjin commented on js-api-doc/exception.d.ts:
```diff
@@ -1,13 +1,41 @@
+   */
   sassStack: string;
 
+  /** The location the error ocurred in the Sass file that triggered it. */
```
```suggestion
  /** The location the error occurred in the Sass file that triggered it. */
```

Awjin commented on js-api-doc/value/number.d.ts:
```diff
@@ -2,83 +2,308 @@ import {List} from 'immutable';
 
+  /**
+   * Returns [[value]] if it's within `min` and `max`, or if it's equal to `min`
+   * or `max` according to Sass's equality returns [[value]] clamped to that
```
nit: The wording here is a bit confusing

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+   * passed to [[compileAsync]] and [[compileStringAsync]], not [[compile]] or
+   * [[compileString]].
+   *
+   * @throws any - If this importer recognizes the URL but determines that
```
Here and below, why is this typed as `any` vs `Error`?

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+   *   - `foo/bar/baz/_index.sass`
+   *   - `foo/bar/baz/index.sass`
+   *   - `foo/bar/baz/_index.scss`
+   *   - `foo/bar/baz/index.scss`
```
Nit: Is order meaningful here? If not consistency with previous example could be nice:

```suggestion
   *   - `foo/bar/baz/index.sass`
   *   - `foo/bar/baz/index.scss`
   *   - `foo/bar/baz/_index.sass`
   *   - `foo/bar/baz/_index.scss`
```

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+ * @category Importer
+ */
 export interface Importer<sync extends 'sync' | 'async' = 'sync' | 'async'> {
+  /**
```
Not a review comment per se, but I really liked how clear this section is!

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+   * be wrapped by Sass. If the exception object has a `message` property, it
+   * will be used as the wrapped exception's message; otherwise, the exception
+   * object's `toString()` will be used. This means it's safe for importers to
+   * throw plain strings.
```
Ah... this answers my earlier question about throwing `any`. Might be worth duplicating this in the other jsdocs.

nex3 replied:
Good thought, I've updated all three to use `@throws` annotations with more detail.

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+   * throw plain strings.
+   *
+   * @param canonicalUrl - The canonical URL of the stylesheet to load. This is
+   * guaranteed to come froma call to [[canonicalize]], although not every call
```
```suggestion
   * guaranteed to come from a call to [[canonicalize]], although not every call
```

Awjin commented on js-api-doc/importer.d.ts:
```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
   load(canonicalUrl: URL): PromiseOr<ImporterResult | null, sync>;
 
+  /** @hidden */
   findFileUrl?: never;
```
Could you clarify why this field exists? (same with `FileImporter.canonicalize`)

stof replied:
This make the type checker complain if you pass an object implementing both `canonicalize` and `findFileUrl` (it is forbidden to try being both a FileImporter and an Importer).

Awjin commented on js-api-doc/options.d.ts:
```diff
@@ -5,44 +5,403 @@ import {Logger} from './logger';
+ * if it is the function may only be passed to [[compileAsync]] and
+ * [[compileStringAsync]], not [[compile]] or [[compileString]].
+ *
+ * @throws any - This function may throw an error, which the Sass compiler will
```
Why not `Error`?

Awjin commented on js-api-doc/options.d.ts:
```diff
@@ -5,44 +5,403 @@ import {Logger} from './logger';
+   *   [[SassNumber.assertInt]], which should be used when possible.
+   *
+   * * In Sass, every value counts as a list. Rather than trying to detect the
+   *   [[SassList]] type, and should use [[Value.asList]] to treat all values as
```
```suggestion
   *   [[SassList]] type, you should use [[Value.asList]] to treat all values as
```

Awjin commented on js-api-doc/options.d.ts:
```diff
@@ -5,44 +5,403 @@ import {Logger} from './logger';
+   *   is represented in UTF-16 as two code units (`0xD83D` and `0xDE0A`). So in
+   *   JavaScript, `"a😊b".charCodeAt(1)` returns `0xD83D`, whereas in Sass
+   *   `str-slice("a😊b", 1, 1)` returns `"😊"`. Functions should follow Sass's
+   *   convention. The [[SassString.sassIndexToStringIndex]]can be used to do
```
```suggestion
   *   convention. [[SassString.sassIndexToStringIndex]] can be used to do
```

Awjin commented on js-api-doc/options.d.ts:
```diff
@@ -5,44 +5,403 @@ import {Logger} from './logger';
+
+  /**
+   * The canonical URL of the entrypoint stylesheet. If this is passed along
+   * with [[impoter]], it's used to resolve relative loads in the entrypoint
```
```suggestion
   * with [[importer]], it's used to resolve relative loads in the entrypoint
```

Awjin commented on js-api-doc/compile.d.ts:
````diff
@@ -3,28 +3,150 @@ import {RawSourceMap} from 'source-map-js';
+ * @example
+ *
+ * ```js
+ * const sass = require('sass');
````
nit: Should we use ESM here and throughout?

```suggestion
 * import * as sass from 'sass';
```

nex3 replied:
Since afaik the latest version of Node still chokes on this without special configuration, I think it's not yet time to do it by default. People who are using ESM will know how to translate better than unfamiliar people. At some point we should make the switch, though.

stof commented on spec/js-api/options.d.ts:
```diff
@@ -110,23 +110,28 @@ export interface Options<sync extends 'sync' | 'async'> {
 
-/** Additional options specific to compiling from a string. */
-export type StringOptions<sync extends 'sync' | 'async'> = Options<sync> & {
+export interface StringOptionsWithoutImporter<sync extends 'sync' | 'async'>
```
should this type have `importer?: never`  to ensure that passing an `importer` option gets typed as `StringOptionsWithImporter` and reports a missing `url` if needed.

nex3 replied:
It turns out this doesn't work because `StringOptionsWithImporter` extends `StringOptionsWithoutImporter`. But that superclass relationship should provide pretty decent error messages on its own.

nex3 committed:
Add .get() utility methods for lists and maps

nex3 committed:
Make the new JS API a bit more idiomatic

This takes more advantage of overloads rather than using factory
constructors or methods with different names.

nex3 committed:
Document the new JS API

nex3 committed:
Merge remote-tracking branch 'origin/main' into new-api-docs

nex3 committed:
Update expected release version of the new JS API
